src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,304 @@
+
+                if (context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeCompilerServicesRuntimeFeatureVirtualStaticsInInterfaces, out INamedTypeSymbol? runtimeFeatureTypeVirtualStaticsInInterfaces))
+                {
+                    ProcessPreviewAttribute(runtimeFeatureTypeVirtualStaticsInInterfaces, requiresPreviewFeaturesSymbols);
+                }
+
+                // Handle user side invocation/references to preview features
+                context.RegisterOperationAction(context => BuildSymbolInformationFromOperations(context, requiresPreviewFeaturesSymbols),
+                    OperationKind.Invocation,
+                    OperationKind.ObjectCreation,
+                    OperationKind.PropertyReference,
+                    OperationKind.FieldReference,
+                    OperationKind.DelegateCreation,
+                    OperationKind.EventReference,
+                    OperationKind.Unary,
+                    OperationKind.Binary,
+                    OperationKind.ArrayCreation,
+                    OperationKind.CatchClause
+                    );
+
+                // Handle library side definitions of preview features
+                context.RegisterSymbolAction(context => AnalyzeSymbol(context, requiresPreviewFeaturesSymbols, requiresPreviewFeaturesSymbolsToUsageType, runtimeFeatureTypeVirtualStaticsInInterfaces), s_symbols);
+            });
+        }
+
+        private static bool TypeSymbolHasAttribute(ITypeSymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType)
+        {
+            ImmutableArray<INamedTypeSymbol> interfaces = symbol.Interfaces;
+            foreach (INamedTypeSymbol? anInterface in interfaces)
+            {
+                var interfaceMembers = anInterface.GetMembers();
+                if (interfaceMembers.Length == 0)
+                {
+                    // Only tag empty interfaces to prevent breaking changes in the future
+                    requiresPreviewFeaturesSymbolsToUsageType.GetOrAdd(symbol, PreviewFeatureUsageType.EmptyInterface);
+                    return true;
+                }
+            }
+
+            INamedTypeSymbol? baseType = symbol.BaseType;
+            if (baseType != null)
+            {
+                return ProcessPreviewAttribute(baseType, requiresPreviewFeaturesSymbols);
+            }
+
+            return false;
+        }
+
+        private static bool ContainingTypeIsPreview(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols)
+        {
+            INamedTypeSymbol? containingType = symbol.ContainingType;
+            // Namespaces do not have attributes
+            while (containingType is INamespaceSymbol)
+            {
+                containingType = containingType.ContainingType;
+            }
+            if (containingType != null)
```

Comment:
Nit: New line between compound statements.
```suggestion

            if (containingType != null)
```